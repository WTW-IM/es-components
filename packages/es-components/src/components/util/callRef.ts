import React from 'react';

const callRef = <T, R extends React.ForwardedRef<T>>(ref: R, value: T) =>
  typeof ref === 'function' ? ref(value) : ref && (ref.current = value);

export default callRef;
