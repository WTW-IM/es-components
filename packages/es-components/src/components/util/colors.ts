import tinycolor, { ColorInput } from 'tinycolor2';

export function darken(color: ColorInput, percent: number) {
  return tinycolor(color).darken(percent).toHexString();
}

export function lighten(color: ColorInput, percent: number) {
  return tinycolor(color).lighten(percent).toHexString();
}

export function getTextColor(
  bgColor: ColorInput,
  darkestColor: ColorInput = 'black',
  lightestColor: ColorInput = 'white'
) {
  return tinycolor.readability(darkestColor, bgColor) > 10
    ? darkestColor
    : lightestColor;
}
