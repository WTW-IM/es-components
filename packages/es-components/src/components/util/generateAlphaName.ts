export default function generateAlphaName(): string {
  const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
  let text = '';
  for (let i = 0; i < 6; i++) {
    text += possible.charAt(Math.floor(Math.random() * possible.length));
  }
  return text;
}
