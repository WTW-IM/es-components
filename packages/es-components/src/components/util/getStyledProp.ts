import { DefaultTheme, ThemeProps } from 'styled-components';
type StringKeys<O> = Extract<keyof O, string>;
type StringKey<O, T extends StringKeys<O>> = T extends StringKeys<O>
  ? T
  : never;
type PathBlock<O, T extends string> = T extends StringKeys<O> ? T : never;
type PathPeriod<O, T extends StringKeys<O>> = `${PathBlock<O, T>}.`;
type PathSegment<
  O,
  T extends string,
  F extends string = T
> = T extends `${PathPeriod<O, infer P>}${infer R}`
  ? PathSegment<O[P], R, F>
  : T extends PathBlock<O, T>
  ? F
  : never;
type PathValue<O, T extends string> = T extends `${PathPeriod<
  O,
  infer P
>}${infer R}`
  ? PathValue<O[P], R>
  : T extends keyof O
  ? O[T]
  : never;

export type StyledPropPathValue<
  O,
  PPath,
  TKey = StringKeys<O>
> = TKey extends StringKey<O, infer T>
  ? PPath extends PathSegment<O[T], infer P>
    ? PathValue<O[T], P>
    : PPath extends PathSegment<O[StringKeys<O>], infer P>
    ? PathValue<O[StringKeys<O>], P>
    : never
  : never;

export type ESThemeProps = ThemeProps<DefaultTheme>;

export function getProp<
  O extends ESThemeProps,
  P extends string,
  T extends StringKeys<O>
>(propPath: P, propTarget: T, props: O): StyledPropPathValue<O, P, T> {
  let targ: unknown = props[propTarget];
  const properties = propPath.split('.');
  while (properties.length && targ) {
    targ = (targ as { [key: string]: unknown })[properties.shift() as string];
  }
  return targ as StyledPropPathValue<O, P, T>;
}

export type InputStylesHeight = StyledPropPathValue<
  ESThemeProps,
  'inputStyles.inputHeight',
  'theme'
>;

export function getStyledProp<P extends string>(
  propPath: P
): (props: ESThemeProps) => StyledPropPathValue<ESThemeProps, P, 'theme'>;
export function getStyledProp<P extends string>(
  propPath: P,
  givenProps: ESThemeProps
): StyledPropPathValue<ESThemeProps, P, 'theme'>;
export function getStyledProp<
  P extends string,
  T extends StringKeys<O>,
  O extends ESThemeProps = ESThemeProps
>(propPath: P, propTarget: T, givenProps: O): StyledPropPathValue<O, P, T>;
export default function getStyledProp<
  P extends string,
  T extends StringKeys<O>,
  O extends ESThemeProps = ESThemeProps
>(
  propPath: P,
  propTarget?: T | O | undefined,
  givenProps?: O | undefined
): ((props: O) => StyledPropPathValue<O, P, T>) | StyledPropPathValue<O, P, T> {
  const themeProp = 'theme' as Extract<keyof O, string>;
  if (!propTarget && !givenProps) {
    return (props: O) => getProp(propPath, themeProp, props);
  }

  if (!givenProps) {
    return getProp(propPath, themeProp, propTarget as O);
  }

  return getProp(
    propPath,
    (propTarget as T | undefined) || themeProp,
    givenProps
  );
}
