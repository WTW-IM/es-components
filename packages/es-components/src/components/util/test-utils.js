// eslint-env jest
import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render } from '@testing-library/react';

export function ThemeComponent(props) {
  return <ThemeProvider theme={viaTheme} {...props} />;
}

ThemeComponent.propTypes = {
  children: PropTypes.node
};

export function renderWithTheme(component) {
  return render(<ThemeComponent>{component}</ThemeComponent>);
}
