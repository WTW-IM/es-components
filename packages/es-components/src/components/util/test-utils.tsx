import React, { Profiler } from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render, RenderResult } from '@testing-library/react';

export function ThemeComponent(
  props: Omit<React.ComponentPropsWithoutRef<typeof ThemeProvider>, 'theme'>
) {
  return <ThemeProvider {...props} theme={viaTheme} />;
}

ThemeComponent.propTypes = {
  children: PropTypes.node
};

const profilerRender = (id: string, phase: string, actualDuration: number) => {
  if (actualDuration < 100) return;
  console.warn('Render took too long!', { id, phase, actualDuration });
};

export function renderWithTheme(component: React.ReactElement): RenderResult {
  const { ...viewUtils } = render(<ThemeComponent>{component}</ThemeComponent>);
  return {
    ...viewUtils,
    rerender: (rerenderComponent: React.ReactNode) =>
      viewUtils.rerender(
        <Profiler id="ThemeRender" onRender={profilerRender}>
          <ThemeComponent>{rerenderComponent}</ThemeComponent>
        </Profiler>
      )
  };
}

export const setClientWidth = (width: number) => {
  Object.defineProperty(document.documentElement, 'clientWidth', {
    writable: true,
    configurable: true,
    value: width
  });
};

export const setClientHeight = (height: number) => {
  Object.defineProperty(document.documentElement, 'clientHeight', {
    writable: true,
    configurable: true,
    value: height
  });
};
