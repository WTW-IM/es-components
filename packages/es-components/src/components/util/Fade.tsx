import React, { useRef, forwardRef } from 'react';
import PropTypes from 'prop-types';
import Transition, {
  TransitionStatus,
  TimeoutProps,
  TransitionProps
} from 'react-transition-group/Transition';
import { callRefs } from './callRef';

type Opacity = React.CSSProperties['opacity'];

export interface FadeProps extends Partial<TimeoutProps<HTMLElement>> {
  duration?: number;
  opacity?: Opacity;
  children: NonNullable<React.ReactNode>;
}

const getFadeStyle = (status: TransitionStatus, finalOpacity: Opacity) => {
  switch (status) {
    case 'entering':
    case 'entered':
      return { finalOpacity };
    default:
      return { opacity: 0 };
  }
};

type AnyReactHTMLElement = React.ReactHTMLElement<HTMLElement>;
type TransitionChildProps = {
  children: AnyReactHTMLElement;
  style: React.CSSProperties;
  transitionStatus: TransitionStatus;
};

const TransitionChild = forwardRef<unknown, TransitionChildProps>(
  function ForwardedTransitionChild(
    { children, style, transitionStatus },
    ref
  ) {
    return React.cloneElement(children, {
      ref: (el: unknown) =>
        callRefs(el, ref, children.ref as React.Ref<unknown> | undefined),
      style: {
        ...((children?.props?.style as Maybe<React.CSSProperties>) || {}),
        ...style,
        ...getFadeStyle(
          transitionStatus,
          typeof style.opacity == 'number' || Boolean(style.opacity)
            ? style.opacity
            : 1
        )
      }
    });
  }
);

type TransitionElement = HTMLElement | undefined;

type FadeTransitionProps<T extends TransitionElement> = Omit<
  TransitionChildProps,
  'transitionStatus'
> &
  Omit<TransitionProps<T>, 'timeout'> & {
    timeout: NonNullable<TransitionProps<T>['timeout']>;
  };

function FadeTransition<T extends TransitionElement>({
  children,
  style,
  ...transitionProps
}: FadeTransitionProps<T>) {
  const nodeRef = useRef(null);
  const child = React.Children.only(children);

  return (
    <Transition nodeRef={nodeRef} {...transitionProps}>
      {transitionStatus => (
        <TransitionChild
          ref={nodeRef}
          transitionStatus={transitionStatus}
          style={style}
        >
          {child}
        </TransitionChild>
      )}
    </Transition>
  );
}

const Fade: React.FC<FadeProps> = ({
  children,
  duration = 150,
  opacity = 1,
  ...otherProps
}) => {
  const transitionStyles = {
    transition: `opacity ${duration || 150}ms linear`,
    opacity
  };

  const transitionProps = {
    ...otherProps,
    timeout: duration || otherProps.timeout || 150
  };

  return React.Children.map(children, child =>
    !React.isValidElement(child) ? (
      child
    ) : (
      <FadeTransition {...transitionProps} style={transitionStyles}>
        {child as AnyReactHTMLElement}
      </FadeTransition>
    )
  );
};

Fade.propTypes = {
  children: PropTypes.any.isRequired,
  duration: PropTypes.number,
  opacity: PropTypes.number
};

export default Fade;
