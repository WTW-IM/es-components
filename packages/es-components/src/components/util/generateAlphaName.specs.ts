import generateAlphaName from './generateAlphaName';

describe('generateAlphaName', () => {
  it('generates a 6 character string', () => {
    const result = generateAlphaName();

    expect(result.length).toBe(6);
  });

  it('generates only alpha characters', () => {
    const result = generateAlphaName();

    expect(/^[a-zA-Z]+$/.test(result)).toBe(true);
  });

  it('generates a unique string each subsequent execution', () => {
    const result1 = generateAlphaName();
    const result2 = generateAlphaName();

    expect(result1).not.toEqual(result2);
  });
});
