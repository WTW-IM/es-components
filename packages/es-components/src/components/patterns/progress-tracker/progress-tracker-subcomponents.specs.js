import React from 'react';
import { waitFor, screen } from '@testing-library/react';
import viaTheme from 'es-components-via-theme';

jest.mock('../../util/useUniqueId', () => ({
  __esModule: true,
  default: jest.fn(() => '123')
}));

import {
  ProgressContainer,
  ProgressItem
} from './progress-tracker-subcomponents';
import { renderWithTheme } from '../../util/test-utils';
import sizes from './pt-sizes';

const { ACTIVE, INACTIVE } = sizes;

test.each([
  { numberOfSteps: 2, percentage: 100 },
  { numberOfSteps: 3, percentage: 50 },
  { numberOfSteps: 4, percentage: (1 / 3) * 100 },
  { numberOfSteps: 5, percentage: 25 }
])(
  'with $numberOfSteps steps, breaks the background line at $percentage%',
  ({ numberOfSteps, percentage }) => {
    const { gray9, gray5 } = viaTheme.colors;
    renderWithTheme(
      <ProgressContainer
        data-testid="progress-tracker"
        activeStepIndex={1}
        numberOfSteps={numberOfSteps}
      />
    );

    expect(screen.getByTestId('progress-tracker')).toHaveStyle(`
      background-image: linear-gradient(
          to right,
          ${gray9} 0%,
          ${gray9} ${percentage}%,
          ${gray5} ${percentage}%,
          ${gray5} 100%)
    `);
  }
);

it('sets the appropriate margin-top when inactive', async () => {
  renderWithTheme(
    <ProgressItem active={false} isPastStep={false} numberOfSteps={2} />
  );
  const topMargin = ACTIVE / 2 - INACTIVE / 2;

  const instance = await screen.findByRole('button');
  expect(instance).toHaveStyle(`
    margin-top: ${topMargin}px;
  `);
  expect(instance).toHaveStyleRule('margin-top', '0', {
    media: `(min-width:${viaTheme.screenSize.tablet})`
  });
});

it('sets the appropriate margin-top when active', async () => {
  renderWithTheme(
    <ProgressItem active={true} isPastStep={false} numberOfSteps={2} />
  );

  const instance = await screen.findByRole('button');

  await waitFor(() => {
    expect(instance).toHaveStyle('margin-top: 0');
  });
  expect(instance).toHaveStyleRule('margin-top', '0', {
    media: `(min-width:${viaTheme.screenSize.tablet})`
  });
});

it('sets the appropriate styles for isPastStep', () => {
  renderWithTheme(
    <ProgressItem active={false} isPastStep={true} numberOfSteps={2} />
  );

  const instance = screen.getByRole('button');
  expect(instance).toMatchSnapshot();
});

it('cannot be clicked when not isPastStep', async () => {
  const clicked = jest.fn();
  renderWithTheme(
    <ProgressItem
      active={false}
      isPastStep={false}
      numberOfSteps={2}
      onPastStepClicked={clicked}
    />
  );
  const instance = await screen.findByRole('button');

  instance.click();
  await waitFor(() => {
    expect(clicked).not.toHaveBeenCalled();
  });
});

it('can be clicked when isPastStep', async () => {
  const clicked = jest.fn();
  renderWithTheme(
    <ProgressItem
      active={false}
      isPastStep={true}
      numberOfSteps={2}
      onPastStepClicked={clicked}
    />
  );

  const instance = await screen.findByRole('button');
  instance.click();
  await waitFor(() => {
    expect(clicked).toHaveBeenCalled();
  });
});
