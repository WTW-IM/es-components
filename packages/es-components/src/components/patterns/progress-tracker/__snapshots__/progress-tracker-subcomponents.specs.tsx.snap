// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets the appropriate styles for isPastStep 1`] = `
.c0 {
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
  padding: 0;
  border: none;
  margin-top: 12.5px;
  background: none;
  counter-increment: progress-tracker-counter;
  font-family: 'Source Sans Pro',sans-serif;
  text-align: center;
}

.c0:active,
.c0:hover,
.c0:focus {
  outline: none;
}

.c0 span {
  display: none;
  color: #444;
  font-size: 18px;
  line-height: 1.1em;
}

.c0::before {
  position: relative;
  display: block;
  width: 20px;
  height: 20px;
  box-sizing: border-box;
  border: 1px solid #c5c5c5;
  border-radius: 20px;
  margin-bottom: 5px;
  background-color: white;
  box-shadow: 0 0 0 3px white;
  color: #444;
  content: counter(progress-tracker-counter);
  font-size: 16px;
  line-height: 1em;
  text-align: center;
}

.c1 {
  cursor: pointer;
}

.c1:hover::before {
  box-shadow: rgba(0, 0, 0, 0.35) 0 0 3px 1px;
}

.c1:active span,
.c1:hover span,
.c1:focus span {
  text-decoration: underline;
}

.c1:focus::before {
  box-shadow: #83bffc 0 0 4px 2px;
}

.c1:focus span {
  outline: 1px dotted #83bffc;
}

.c1::before {
  border-color: #444;
}

@media (min-width: 768px) {
  .c0 {
    margin: 0 auto;
    margin-top: 0;
    font-size: 18px;
  }
}

@media (min-width: 768px) {
  .c0 span {
    display: inline;
    font-weight: normal;
  }
}

@media (min-width: 768px) {
  .c0::before {
    width: 41px;
    height: 41px;
    border-width: 3px;
    border-radius: 41px;
    font-size: 20.008px;
    font-weight: normal;
    line-height: 1.7em;
  }
}

@media (min-width: 768px) {
  .c1::before {
    border-width: 5px;
    font-weight: bold;
    line-height: 1.5em;
  }
}

@media (min-width: 768px) {
  .c1 span {
    font-weight: bold;
  }
}

@media (min-width: 768px) {

}

<button
  aria-labelledby="123-span"
  class="c0 c1"
  id="123"
>
  <span
    id="123-span"
  />
</button>
`;
